include/hydrosheds/hydrosheds.hh

```diff
@@ -4,59 +4,193 @@
-    // What about upstream segments - is it possible?
 
     private:
-    friend class HydroshedsDataSet;
+        friend class HydroshedsDataSet;
+        friend class FullDatasetRiverSegmentIterator;
+        friend class DownstreamIteratorHelper;
+        friend class DownstreamIterator;
+
+        RiverSegment(OGRLayer*, OGRFeature*);
+        OGRLayer* layer;
+        OGRFeature* feature;
+    };
+
+
+    class FullDatasetRiverSegmentIterator
+    {
+    private:
+        OGRFeature *feature;
+        OGRLayer *layer;
+        RiverSegment segment;
+    public:
+
+        /** @brief Constructs an iterator with a given OGRFeature and OGRLayer
+        *
+         *   This Constructor initializes the segment with the two given parameters
+         *
+         *
+        *   @param ogrFeature that this iterator should be validating against
+        *  @param ogrLayer that this layer should be validating the layer
+        */
+        FullDatasetRiverSegmentIterator(OGRFeature* ogrFeature, OGRLayer* ogrLayer);
+
+
+
+        /** @brief Points to the segment of the FullDatasetRiverSegmentIterator
+         *
+         *
+         * @returns a pointer to the segment instances form the Itertor instacne
+         */
+        RiverSegment* operator->();
+
+
+        /** @brief A reference to the segment of the FullDatasetRiverSegmentIterator
+         *
+         * @return a reference to reference to the segment
+         */
+        RiverSegment& operator*();
+
+
+
+        /** @brief Prefix incrementation
+         *
+         *  This increment operator increments the iterator by assiging its feature to
+         *  the next upcoming feature defined by the layer. It will be incremented first
+         *  and then returned
+         *
+         * @returns an incremented version of the iterator
+         */
+        FullDatasetRiverSegmentIterator operator++();
+
+
+        /** @brief Postfix incrementation
+         *
+         * This increment operator returns the current Iterator first and
+         * then increments the Iterator
+         *
+         * @return an incremented version of the iterator
+         */
+        FullDatasetRiverSegmentIterator operator++(int);
+
+        /** @brief comparison operator
+         *
+         * @param a an Iterator from the class FullDatasetRiverSegmentIterator
+         * @returns whether or not this Iterator is equal to a or not
+         */
+        bool operator==(const FullDatasetRiverSegmentIterator& a);
+        // equality operations
+
+        /** @brief comparison operator
+         *
+         * @param a an Iterator from the class FullDatasetRiverSegmentIterator
+         * @returns whether or not this Iterator is equal to a or not
+         */
+        bool operator!=(const FullDatasetRiverSegmentIterator& a);
+
+
+    };
+
+
+
+    namespace impl{
+        double norm(hydrosheds::Coordinate x, hydrosheds::Coordinate y);
+
+        struct DownstreamIterator{
```

Comment:
I don't think the `DownstreamIterator` should be in the `impl` namespace, as users will see this type.